In [1]:
# Update sklearn to prevent version mismatches
!pip install sklearn --upgrade

Requirement already up-to-date: sklearn in c:\users\cgotb\anaconda3\lib\site-packages (0.0)


In [2]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
!pip install joblib

In [3]:
#create dataframes, use numpy to display process
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#import tensorflow as tf
#from keras.models import Sequential
#from keras.utils import to_categorical
#from keras.layers import Dense
#from sklearn.preprocessing import LabelEncoder
#from keras.callbacks import EarlyStopping

# Read the CSV and Perform Basic Data Cleaning

In [4]:
#read csv and do data mungling by dropping/removing nulls
exo_df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
exo_df = exo_df.dropna(axis='columns', how='all')
# Drop the null rows
exo_df = exo_df.dropna()
exo_df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [5]:
exo_df.columns

Index(['koi_disposition', 'koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co',
       'koi_fpflag_ec', 'koi_period', 'koi_period_err1', 'koi_period_err2',
       'koi_time0bk', 'koi_time0bk_err1', 'koi_time0bk_err2', 'koi_impact',
       'koi_impact_err1', 'koi_impact_err2', 'koi_duration',
       'koi_duration_err1', 'koi_duration_err2', 'koi_depth', 'koi_depth_err1',
       'koi_depth_err2', 'koi_prad', 'koi_prad_err1', 'koi_prad_err2',
       'koi_teq', 'koi_insol', 'koi_insol_err1', 'koi_insol_err2',
       'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_steff_err1',
       'koi_steff_err2', 'koi_slogg', 'koi_slogg_err1', 'koi_slogg_err2',
       'koi_srad', 'koi_srad_err1', 'koi_srad_err2', 'ra', 'dec',
       'koi_kepmag'],
      dtype='object')

# Select your features (columns)

In [6]:
# split x and y values into two variables, (one to train and one to test) (separate)
# we do this to note outliers in data 
X= exo_df[['koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co',
       'koi_fpflag_ec', 'koi_period', 'koi_period_err1', 'koi_period_err2',
       'koi_time0bk', 'koi_time0bk_err1', 'koi_time0bk_err2', 'koi_impact',
       'koi_impact_err1', 'koi_impact_err2', 'koi_duration',
       'koi_duration_err1', 'koi_duration_err2', 'koi_depth', 'koi_depth_err1',
       'koi_depth_err2', 'koi_prad', 'koi_prad_err1', 'koi_prad_err2',
       'koi_teq', 'koi_insol', 'koi_insol_err1', 'koi_insol_err2',
       'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_steff_err1',
       'koi_steff_err2', 'koi_slogg', 'koi_slogg_err1', 'koi_slogg_err2',
       'koi_srad', 'koi_srad_err1', 'koi_srad_err2', 'ra', 'dec',
       'koi_kepmag']]
y = exo_df['koi_disposition']

In [7]:
X.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
0              0              0              0              0   54.418383   
1              0              1              0              0   19.899140   
2              0              1              0              0    1.736952   
3              0              0              0              0    2.525592   
4              0              0              0              0    4.134435   

   koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
0     2.479000e-04    -2.479000e-04   162.513840          0.003520   
1     1.490000e-05    -1.490000e-05   175.850252          0.000581   
2     2.630000e-07    -2.630000e-07   170.307565          0.000115   
3     3.760000e-06    -3.760000e-06   171.595550          0.001130   
4     1.050000e-05    -1.050000e-05   172.979370          0.001900   

   koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0         -0.003520  ...             -81      4.467           0.064   
1         -0.000581  ...            -176      4.544           0.044   
2         -0.000115  ...            -174      4.564           0.053   
3         -0.001130  ...            -211      4.438           0.070   
4         -0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 40 columns]

In [8]:
X.columns

Index(['koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co', 'koi_fpflag_ec',
       'koi_period', 'koi_period_err1', 'koi_period_err2', 'koi_time0bk',
       'koi_time0bk_err1', 'koi_time0bk_err2', 'koi_impact', 'koi_impact_err1',
       'koi_impact_err2', 'koi_duration', 'koi_duration_err1',
       'koi_duration_err2', 'koi_depth', 'koi_depth_err1', 'koi_depth_err2',
       'koi_prad', 'koi_prad_err1', 'koi_prad_err2', 'koi_teq', 'koi_insol',
       'koi_insol_err1', 'koi_insol_err2', 'koi_model_snr', 'koi_tce_plnt_num',
       'koi_steff', 'koi_steff_err1', 'koi_steff_err2', 'koi_slogg',
       'koi_slogg_err1', 'koi_slogg_err2', 'koi_srad', 'koi_srad_err1',
       'koi_srad_err2', 'ra', 'dec', 'koi_kepmag'],
      dtype='object')

# Create a Train Test Split

Use `koi_disposition` for the y values

In [9]:
#train split/separated data
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, random_state=1, train_size=0.5)
x_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
5390              1              0              0              0    3.343285   
3301              0              0              0              0    0.876850   
4528              0              0              0              0  143.163697   
6430              0              1              1              0    5.462689   
5378              0              1              0              0    2.252823   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
5390     4.380000e-05    -4.380000e-05   134.845100          0.011200   
3301     7.720000e-06    -7.720000e-06   131.770330          0.008320   
4528     2.400000e-03    -2.400000e-03   243.713300          0.011700   
6430     6.540000e-06    -6.540000e-06   134.265160          0.000971   
5378     1.340000e-07    -1.340000e-07   131.633146          0.000049   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
5390         -0.011200  ...            -197      3.773           0.293   
3301         -0.008320  ...             -85      4.069           0.174   
4528         -0.011700  ...            -254      3.050           0.030   
6430         -0.000971  ...            -159      4.479           0.078   
5378         -0.000049  ...            -182      4.431           0.105   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
5390          -0.098     2.652          0.433         -0.939  296.86258   
3301          -0.116     1.756          0.295         -0.406  297.19199   
4528          -0.031     7.653          0.334         -2.023  297.78833   
6430          -0.182     0.888          0.238         -0.102  295.45105   
5378          -0.180     0.951          0.248         -0.134  294.97415   

            dec  koi_kepmag  
5390  41.147419      13.276  
3301  40.387951      14.878  
4528  40.130909      12.644  
6430  40.175652      13.266  
5378  48.944580      13.387  

[5 rows x 40 columns]

In [10]:
#train split/separated data
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, random_state=1, train_size=0.5)
x_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
5390              1              0              0              0    3.343285   
3301              0              0              0              0    0.876850   
4528              0              0              0              0  143.163697   
6430              0              1              1              0    5.462689   
5378              0              1              0              0    2.252823   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
5390     4.380000e-05    -4.380000e-05   134.845100          0.011200   
3301     7.720000e-06    -7.720000e-06   131.770330          0.008320   
4528     2.400000e-03    -2.400000e-03   243.713300          0.011700   
6430     6.540000e-06    -6.540000e-06   134.265160          0.000971   
5378     1.340000e-07    -1.340000e-07   131.633146          0.000049   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
5390         -0.011200  ...            -197      3.773           0.293   
3301         -0.008320  ...             -85      4.069           0.174   
4528         -0.011700  ...            -254      3.050           0.030   
6430         -0.000971  ...            -159      4.479           0.078   
5378         -0.000049  ...            -182      4.431           0.105   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
5390          -0.098     2.652          0.433         -0.939  296.86258   
3301          -0.116     1.756          0.295         -0.406  297.19199   
4528          -0.031     7.653          0.334         -2.023  297.78833   
6430          -0.182     0.888          0.238         -0.102  295.45105   
5378          -0.180     0.951          0.248         -0.134  294.97415   

            dec  koi_kepmag  
5390  41.147419      13.276  
3301  40.387951      14.878  
4528  40.130909      12.644  
6430  40.175652      13.266  
5378  48.944580      13.387  

[5 rows x 40 columns]

In [11]:
y_train.head()

5390    FALSE POSITIVE
3301         CANDIDATE
4528         CANDIDATE
6430    FALSE POSITIVE
5378    FALSE POSITIVE
Name: koi_disposition, dtype: object

# Pre-processing

Scale the data using the MinMaxScaler and perform some feature selection

In [12]:
# Scale your data to prep for log regression#
#create a standard Scater model and fit it to the training data
# we do this to note trends/ outliers in data 

from sklearn.preprocessing import MinMaxScaler
x_scaler = MinMaxScaler().fit(x_train)

In [13]:
#transform the training data using the x_scaler variable created above
x_train_scaled = x_scaler.transform(x_train)
x_test_scaled = x_scaler.transform(x_test)

In [14]:
x_train_scaled.shape

(3495, 40)

In [15]:
x_test_scaled.shape

(3496, 40)

# Model 1 - Logistical Regression
* Logistic Regression is a statistical method for predicting binary outcomes from data; e.g. "yes" vs "no" or "young" vs "old". These are categories that translate to probability of being a 0 or a 1.  Calculate logistic regression by adding an activation function as the final step to our linear model. This converts the linear regression output to a probability.

Set up Logistical Regression Model

* Train Model by fitting the y-trained target values and x-trained scaled values to model


* Score Logistical Regression Model by Train and Test - Train Score: 0.791 --- Test Score: 0.821. Logistical Regression model is able to predict an exoplanet based on the features selected from the dataset (not transit like flag, stellar eclipse flag, centroid offset, orbital period, transit duration, transit depth, planetary radius, equilibrium temperature, insolation flux, and other elements) approximately 82% of the time.
* Use grid search to fine tune the parameters for the model and received a best score of 0.784 which is slightly lower than the test score value. Adjusting the C and gamma values may render a different result

In [16]:
# For 1st Model create a logistic regression Model
#increase max_iter
#create classifier 
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(max_iter=1000)
classifier

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

# Train the Model



In [17]:
# Fit (train) or model using the training data
#Train 1st model by fitting the trained x and y values to model, x is scaled to transform to readable value per MC
classifier.fit(x_train_scaled, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [18]:
# Validate the model using the test data
print(f"Training Data Score: {classifier.score(x_train_scaled, y_train)}")
print(f"Testing Data Score: {classifier.score(x_test_scaled, y_test)}")

Training Data Score: 0.8472103004291845
Testing Data Score: 0.8721395881006865


# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [19]:
#Check score parameter to determine if it's within acceptable range,
#see class activity

In [20]:
# Support vector machine linear classifier
from sklearn.svm import SVC 
model = SVC(kernel="linear")

In [21]:
# Create the GridSearchCV model to find tune parameters for model and improve current score. Adjust C and gamma values to see if any significant changes occur. 
from sklearn.model_selection import GridSearchCV
param_grid = {"C": [1, 5, 10],
              "gamma": [0.0001, 0.001, 0.01]}
grid = GridSearchCV(model, param_grid, verbose=3)

In [22]:
# Train the model with GridSearch
# Fit the model using the grid search estimator. 
# This will take the SVC model and try each combination of parameters
#run model
grid.fit(x_train_scaled, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................... C=1, gamma=0.0001, score=0.820, total=   0.4s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.3s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.828, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.6s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.825, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.843, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.823, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.820, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.828, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.825, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.843, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:   14.9s finished


GridSearchCV(cv=None, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='linear', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': [1, 5, 10], 'gamma': [0.0001, 0.001, 0.01]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=3)

In [23]:
#display updated results for parameters and score
print(grid.best_params_)
print(grid.best_score_)

{'C': 10, 'gamma': 0.0001}
0.8652360515021458


# Save the Model

In [24]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = "Logistical Regression Model.sav"
joblib.dump(classifier, filename)

['Logistical Regression Model.sav']